core-bundle/src/Image/Preview/PreviewFactory.php

```diff
@@ -0,0 +1,532 @@
+
+        while ($newSize < $size) {
+            $newSize *= 2;
+        }
+
+        return min($newSize, $this->maxSize);
+    }
+
+    private function getCachedPreviews(string $targetPath, int $firstPage, int $lastPage): ?array
+    {
+        $globPattern = preg_replace('/[*?[{\\\\]/', '\\\\$0', $targetPath).'*.*';
+        $filesFound = [];
+
+        foreach (glob($globPattern) as $cacheFile) {
+            if (\in_array(pathinfo($cacheFile, PATHINFO_EXTENSION), $this->validImageExtensions, true)) {
+                $filesFound[pathinfo($cacheFile, PATHINFO_FILENAME)] = $cacheFile;
+            }
+        }
+
+        $previews = [];
+        $fileName = pathinfo($targetPath, PATHINFO_BASENAME);
+
+        for ($page = $firstPage; $page <= $lastPage; ++$page) {
+            $cacheFileName = $fileName.$this->getPageSuffix($page);
+
+            if (isset($filesFound[$cacheFileName])) {
+                $previews[] = $filesFound[$cacheFileName];
+                continue;
+            }
+
+            if (!isset($filesFound["$fileName-last"])) {
+                return null;
+            }
+
+            $lastName = pathinfo((new Filesystem())->readlink($filesFound["$fileName-last"]), PATHINFO_FILENAME);
+            $lastFound = $fileName.$this->getPageSuffix($page - 1);
+
+            if ($lastName !== $lastFound) {
+                return null;
+            }
+
+            break;
+        }
+
+        return $previews;
+    }
+
+    private function getHeader(string $path): string
+    {
+        $size = 0;
+
+        foreach ($this->previewProviders as $provider) {
+            $size = max($size, $provider->getFileHeaderSize());
+        }
+
+        return $size > 0 ? file_get_contents($path, false, null, 0, $size) : '';
+    }
+
+    private function getPreviewSizeFromWidthHeightDensities(int $width, int $height, string $densities): int
+    {
+        $widthDescriptor = 0;
+        $scaleFactor = 1;
+
+        foreach (explode(',', $densities) as $density) {
+            if ('w' === substr(trim($density), -1)) {
+                $widthDescriptor = max($widthDescriptor, (int) $density);
+            } else {
+                $scaleFactor = max($scaleFactor, (float) $density);
+            }
+        }
+
+        return (int) round(
+            max(
+                max($width, $height) * $scaleFactor,
+                $widthDescriptor,
+            )
+        );
+    }
+
+    private function createCachePath(string $path, int $size, array $previewOptions): string
+    {
+        ksort($previewOptions);
+
+        $hashData = [
+            Path::makeRelative($path, $this->cacheDir),
+            (string) $size,
+            (string) filemtime($path),
+            ...array_keys($previewOptions),
+            ...array_values($previewOptions),
+        ];
+
+        $hash = hash_hmac('sha256', md5(implode('|', $hashData)), $this->secret, true);
+        $hash = substr(strtr(base64_encode($hash), '+/', '-_'), 0, 11);
```

Comment:
@ausi @leo what do you think about making `11` dynamic to increase security even more? It has to be deterministic of course but if it's based on the secret and the $hash it would increase the number of possibilities even more.

```suggestion
        $length = 11 + hexdec(md5($hash . $this->secret)[0]);
        $hash = substr(strtr(base64_encode($hash), '+/', '-_'), 0, $length);
```

Maybe you have a better idea but this variant would generate lengths between 11 and 27.

Reply:
Wrong Leo!

Reply:
I’m not sure if this is how to calculate the brute force possibilities in this case, but I’ve got a 1/16th chance that the length is 11, this can (maybe?) just be multiplied with the probability of 1/(2^65) which would be less then increasing the length from 11 to 12.

So just increasing the length would be better I think (and less magic).

Reply:
Aaah combinatorics :) Wasn't my strength but imho this is permutation (order matters) with repetition (every hex char can appear multiple times). So that would mean

Fixed to 11: `16^11 = 1.75 * 10^13`
Fixed to 26: `16^26 = 2.02 * 10^31`

And if we have a range between 11 and 26, that should be the summary of all of those.
So `16^11 + 16^12 + .... + 16^26`. Which would end up being something like `2.16 * 10^31`.

So yeah, we can just increase the length of the hash, making it dynamic does not add a lot more possibilities.

Reply:
I mean it does add a lot more (`0.14 * 10^31` is still a huge number) but compared to the already huge `2.02 ^ 10^31` it doesn't make any difference :D

Reply:
I'd be more comfortable with more than `11`, though :)

Reply:
It's easier to think of 'bits of information' in this case. A shorter input is basically just padded with zeroes - only length matters.

Reply:
I try another go on this: 

Considering case-insensitive file systems, we have basically 38^11 possible values for the hash. Brute forcing 1000 URLs per second would take around 7.5 million years to try out every hash or an average of 3.8 million years to crack one single hash.


| Chars | Million years max | Million years average |
|-------|------------------:|----------------------:|
| 11    |                 8 |                     4 |
| 12    |               287 |                   144 |
| 13    |             10924 |                  5462 |
| 14    |            415111 |                207555 |
| 15    |          15774200 |               7887100 |
| 15    |         599419599 |             299709800 |

Increasing it to 12 should be enough IMO.

Reply:
What we're missing here is that the possibilities alone are not sufficient for this decision. We also have to account for the probability of shared hash information. See https://en.wikipedia.org/wiki/Birthday_problem. 
I don't really know but instead of arguing I would simply pick a number high enough (e.g. 20) so it doesn't matter at all. Saving 8 letters makes no sense imho.

Reply:
> We also have to account for the probability of shared hash information.

The probability of a hash collision only depends on the number of hashed elements. With around 1M images in the system and a length of 64bit (38 values per char = ~6bit * 11 = ~66bit) [the probability](https://stackoverflow.com/a/62667633/3146819) [is approx.](https://www.wolframalpha.com/input/?i=1+-+e%5E%28-k%5E2%2F2%5E%28n%2B1%29%29) 2,7 * 10^-8 or in other words: negligible.

Reply:
> I don't really know but instead of arguing I would simply pick a number high enough (e.g. 20) so it doesn't matter at all.

That’s what I tried to do by increasing the 36 bits of contao/image to ~64 bits.

@contao/developers should we vote on this one?
- 👎 for 11 characters, e.g.: assets/previews/w/foobar-a6B6UBSExR.png
- 👍 for 16 characters, e.g.: assets/previews/w/foobar-a6B6UBSExRhbpJP.png
- 🚀 for 20 characters, e.g.: assets/previews/w/foobar-a6B6UBSExRhbpJPMTY5.png

Reply:
This is currently blocking the PR from being merged. @Toflar You have not voted yet, I think.

Reply:
Just did. I simply don't see why we would want a short hash. It makes absolutely no sense to me.

Reply:
Increased to 16 in 130d645107c72b519dca519cfb184d04a67b1d03